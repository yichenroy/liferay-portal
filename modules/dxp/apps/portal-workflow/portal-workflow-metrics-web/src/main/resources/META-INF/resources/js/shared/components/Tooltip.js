import getCN from 'classnames';
import React from 'react';

const CLASSNAME = 'workflow-tooltip';

export default class Tooltip extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			displayTooltip: false
		};
	}

	hideTooltip() {
		this.setState({displayTooltip: false});
	}

	showTooltip() {
		this.setState({displayTooltip: true});
	}

	render() {
		const {message, position = 'top', width} = this.props;
		const {displayTooltip} = this.state;

		return (
			<span
				className={CLASSNAME}
				onMouseLeave={this.hideTooltip.bind(this)}
			>
				{displayTooltip && (
					<TooltipBase
						position={position}
						title={message}
						visible={displayTooltip}
						width={width}
					/>
				)}
				<span
					className='tooltip-trigger'
					onMouseOver={this.showTooltip.bind(this)}
				>
					{this.props.children}
				</span>
			</span>
		);
	}
}

export class TooltipBase extends React.Component {
	render() {
		const {position, title, visible, width} = this.props;

		const classes = getCN('tooltip', {
			[`clay-tooltip-${position}`]: position,
			[`workflow-tooltip-${position}`]: position,
			['show']: visible
		});

		const classesArrow = getCN('arrow', 'workflow-tooltip-arrow');

		return (
			<div className={classes} style={{width: `${width}px`}}>
				{position !== 'none' && <div className={classesArrow} />}
				<div className='tooltip-inner'>
					<div>{title}</div>
				</div>
			</div>
		);
	}
}
