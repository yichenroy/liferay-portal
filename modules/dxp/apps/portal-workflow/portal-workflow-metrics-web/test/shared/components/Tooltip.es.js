/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Liferay Enterprise
 * Subscription License ("License"). You may not use this file except in
 * compliance with the License. You can obtain a copy of the License by
 * contacting Liferay, Inc. See the License for the specific language governing
 * permissions and limitations under the License, including but not limited to
 * distribution rights of the Software.
 */

import Tooltip, {
	TooltipBase
} from '../../../src/main/resources/META-INF/resources/js/shared/components/Tooltip.es';
import React from 'react';

describe('Tooltip', () => {
	let component;

	const props = {
		message: 'Tooltip Message',
		position: 'top',
		width: '200'
	};

	afterEach(() => {
		if (component) {
			component.unmount();
		}
	});

	it('Should render component with tooltip after mouse over', () => {
		component = shallow(
			<Tooltip {...props} position={'right'}>
				{'Target'}
			</Tooltip>
		);

		component.find('.tooltip-trigger').simulate('mouseover');

		expect(component).toMatchSnapshot();
	});

	it('Should render component with tooltip markup', () => {
		component = shallow(<Tooltip {...props}>{'Target'}</Tooltip>);

		const instance = component.instance();

		instance.showTooltip();

		expect(component).toMatchSnapshot();
	});

	it('Should render component with tooltip with bottom position', () => {
		jest.useFakeTimers();

		component = shallow(
			<Tooltip {...props} position={'bottom'}>
				{'Target'}
			</Tooltip>
		);

		const instance = component.instance();

		instance.showTooltip();

		expect(component.find(TooltipBase).render()).toMatchSnapshot();
	});

	it('Should render component with tooltip with left position', () => {
		component = shallow(
			<Tooltip {...props} position={'right'}>
				{'Target'}
			</Tooltip>
		);

		const instance = component.instance();

		instance.showTooltip();

		expect(component.find(TooltipBase).render()).toMatchSnapshot();
	});

	it('Should render component with tooltip with right position', () => {
		component = shallow(
			<Tooltip {...props} position={'right'}>
				{'Target'}
			</Tooltip>
		);

		const instance = component.instance();

		instance.showTooltip();

		expect(component.find(TooltipBase).render()).toMatchSnapshot();
	});

	it('Should render component with tooltip with top position', () => {
		component = shallow(<Tooltip {...props}>{'Target'}</Tooltip>);

		const instance = component.instance();

		instance.showTooltip();

		expect(component.find(TooltipBase).render()).toMatchSnapshot();
	});

	it('Should render component without tooltip after mouse leave', () => {
		component = shallow(
			<Tooltip {...props} position={'right'}>
				{'Target'}
			</Tooltip>
		);

		component.find('.tooltip-trigger').simulate('mouseover');
		component.find('.workflow-tooltip').simulate('mouseleave');

		expect(component).toMatchSnapshot();
	});

	it('Should render component without tooltip markup', () => {
		component = shallow(<Tooltip {...props}>{'Target'}</Tooltip>);

		expect(component).toMatchSnapshot();
	});
});
