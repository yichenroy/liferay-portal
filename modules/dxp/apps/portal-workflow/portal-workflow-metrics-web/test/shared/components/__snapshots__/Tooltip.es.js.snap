// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Should render component with tooltip after mouse over 1`] = `
<span
  className="workflow-tooltip"
  onMouseLeave={[Function]}
>
  <TooltipBase
    position="right"
    title="Tooltip Message"
    visible={true}
    width="200"
  />
  <span
    className="tooltip-trigger"
    onMouseOver={[Function]}
  >
    Target
  </span>
</span>
`;

exports[`Tooltip Should render component with tooltip markup 1`] = `
<span
  className="workflow-tooltip"
  onMouseLeave={[Function]}
>
  <TooltipBase
    position="top"
    title="Tooltip Message"
    visible={true}
    width="200"
  />
  <span
    className="tooltip-trigger"
    onMouseOver={[Function]}
  >
    Target
  </span>
</span>
`;

exports[`Tooltip Should render component with tooltip with bottom position 1`] = `
<div
  class="tooltip clay-tooltip-bottom workflow-tooltip-bottom show"
  style="width:200px"
>
  <div
    class="arrow workflow-tooltip-arrow"
  />
  <div
    class="tooltip-inner"
  >
    <div>
      Tooltip Message
    </div>
  </div>
</div>
`;

exports[`Tooltip Should render component with tooltip with left position 1`] = `
<div
  class="tooltip clay-tooltip-right workflow-tooltip-right show"
  style="width:200px"
>
  <div
    class="arrow workflow-tooltip-arrow"
  />
  <div
    class="tooltip-inner"
  >
    <div>
      Tooltip Message
    </div>
  </div>
</div>
`;

exports[`Tooltip Should render component with tooltip with right position 1`] = `
<div
  class="tooltip clay-tooltip-right workflow-tooltip-right show"
  style="width:200px"
>
  <div
    class="arrow workflow-tooltip-arrow"
  />
  <div
    class="tooltip-inner"
  >
    <div>
      Tooltip Message
    </div>
  </div>
</div>
`;

exports[`Tooltip Should render component with tooltip with top position 1`] = `
<div
  class="tooltip clay-tooltip-top workflow-tooltip-top show"
  style="width:200px"
>
  <div
    class="arrow workflow-tooltip-arrow"
  />
  <div
    class="tooltip-inner"
  >
    <div>
      Tooltip Message
    </div>
  </div>
</div>
`;

exports[`Tooltip Should render component without tooltip after mouse leave 1`] = `
<span
  className="workflow-tooltip"
  onMouseLeave={[Function]}
>
  <span
    className="tooltip-trigger"
    onMouseOver={[Function]}
  >
    Target
  </span>
</span>
`;

exports[`Tooltip Should render component without tooltip markup 1`] = `
<span
  className="workflow-tooltip"
  onMouseLeave={[Function]}
>
  <span
    className="tooltip-trigger"
    onMouseOver={[Function]}
  >
    Target
  </span>
</span>
`;
